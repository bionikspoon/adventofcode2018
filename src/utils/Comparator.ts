type CompareFn<T> = (l: T, r: T) => 0 | -1 | 1

export default class Comparator<T> {
  private static defaultCompareFunction<T>(l: T, r: T) {
    if (l === r) return 0

    return l < r ? -1 : 1
  }
  private compare: CompareFn<T>
  constructor(fn?: CompareFn<T>) {
    this.compare = fn || Comparator.defaultCompareFunction
  }

  public equal(l: T, r: T) {
    return this.compare(l, r) === 0
  }
  public lessThan(l: T, r: T) {
    return this.compare(l, r) < 0
  }
  public greaterThan(l: T, r: T) {
    return this.compare(l, r) > 0
  }
  public lessThanOrEqual(l: T, r: T) {
    return this.lessThan(l, r) || this.equal(l, r)
  }
  public greaterThanOrEqual(l: T, r: T) {
    return this.greaterThan(l, r) || this.equal(l, r)
  }
  public reverse() {
    const compareOriginal = this.compare
    this.compare = (l, r) => compareOriginal(r, l)
    return this
  }
}
