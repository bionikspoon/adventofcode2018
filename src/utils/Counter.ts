import R from 'ramda'

export default class Counter {
  private counts: { [key: string]: number }

  constructor() {
    this.counts = {}
  }

  public add(value: string | number) {
    const count = this.counts[value] ? this.counts[value] + 1 : 1

    this.counts = { ...this.counts, [value]: count }
    return this
  }

  public deleteValue(v: number) {
    this.counts = R.reject(R.equals(v), this.counts)

    return this
  }

  public getValue(key: string | number) {
    if (!this.counts[key]) return 0

    return this.counts[key]
  }

  public mostCommon() {
    return this.entries().sort(
      ([lKey, lCount], [rKey, rCount]) => rCount - lCount
    )
  }

  public values() {
    return Object.values(this.counts)
  }

  public entries() {
    return Object.entries(this.counts)
  }
}
