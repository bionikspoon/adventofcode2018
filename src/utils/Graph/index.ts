// Borrowed from https://github.com/trekhleb/javascript-algorithms/tree/master/src/data-structures/graph
export { default as Graph } from './Graph'
export { default as GraphEdge } from './GraphEdge'
export { default as GraphVertex } from './GraphVertex'
