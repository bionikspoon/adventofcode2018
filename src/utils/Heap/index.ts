export { default as MinHeap } from './MinHeap'
