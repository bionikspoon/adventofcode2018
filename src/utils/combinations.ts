export default function combinations<T>(items: T[]): Array<[T, T]> {
  return Array.from(iterateCombinations(items))
}

function* iterateCombinations<T>(items: T[]): IterableIterator<[T, T]> {
  for (let i = 0; i < items.length; i++) {
    const I = items[i]
    for (let j = i + 1; j < items.length; j++) {
      const J = items[j]
      yield [I, J]
    }
  }
}
