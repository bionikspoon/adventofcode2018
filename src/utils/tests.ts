import fs from 'fs'
import path from 'path'
import { promisify } from 'util'

const readFileAsync = promisify(fs.readFile)

export const getInput = (dirname: string, file: string) =>
  readFileAsync(path.join(dirname, 'case', file)).then(buffer =>
    buffer.toString()
  )
