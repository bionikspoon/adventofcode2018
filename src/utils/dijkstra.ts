import { Graph, GraphVertex } from './Graph'
import PriorityQueue from './PriorityQueue'

interface IDistances {
  [key: string]: number
}
export interface IDijkstraOptions<T> {
  queueCompareFn?: (
    this: PriorityQueue<T>,
    l: GraphVertex<T>,
    r: GraphVertex<T>
  ) => 0 | -1 | 1
  canTraverse?: (neighbor: GraphVertex<T>) => boolean
  shouldPoll?: (vertex: GraphVertex<T>) => boolean
}
export default function dijkstra<T>(
  graph: Graph<T>,
  startVertex: GraphVertex<T>,
  options: IDijkstraOptions<T> = {}
) {
  const visitedVertices: { [key: string]: GraphVertex<T> } = {}
  const queue = new PriorityQueue<GraphVertex<T>>(options.queueCompareFn)
  const { distances, previousVertices } = initDistances(graph)

  distances[startVertex.getKey()] = 0
  queue.add(startVertex, distances[startVertex.getKey()])

  while (!queue.isEmpty()) {
    const currentVertex = queue.poll()!

    graph.getNeighbors(currentVertex).forEach(neighbor => {
      if (visitedVertices[neighbor.getKey()]) return
      if (options.canTraverse && !options.canTraverse(neighbor)) return

      const distanceToNeighborFromCurrent =
        distances[currentVertex.getKey()] +
        graph.findEdge(currentVertex, neighbor)!.weight

      if (distanceToNeighborFromCurrent < distances[neighbor.getKey()]) {
        distances[neighbor.getKey()] = distanceToNeighborFromCurrent
        updateNeighborPriority(queue, neighbor, distances)
        previousVertices[neighbor.getKey()] = currentVertex
      }

      ensureQueueHasNeighbor(queue, neighbor, distances)
    })

    visitedVertices[currentVertex.getKey()] = currentVertex
    if (options.shouldPoll && !options.shouldPoll(currentVertex)) break
  }

  return { distances, previousVertices }
}
function initDistances<T>(graph: Graph<T>) {
  const distances: IDistances = {}
  const previousVertices: { [key: string]: GraphVertex<T> | null } = {}

  graph.getAllVertices().forEach(vertex => {
    distances[vertex.getKey()] = Infinity
    previousVertices[vertex.getKey()] = null
  })

  return { distances, previousVertices }
}

function updateNeighborPriority<T>(
  queue: PriorityQueue<GraphVertex<T>>,
  neighbor: GraphVertex<T>,
  distances: IDistances
): void {
  if (queue.hasValue(neighbor)) {
    queue.changePriority(neighbor, distances[neighbor.getKey()])
  }
}

function ensureQueueHasNeighbor<T>(
  queue: PriorityQueue<GraphVertex<T>>,
  neighbor: GraphVertex<T>,
  distances: IDistances
): void {
  if (!queue.hasValue(neighbor)) {
    queue.add(neighbor, distances[neighbor.getKey()])
  }
}
