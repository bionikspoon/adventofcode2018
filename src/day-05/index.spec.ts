import { reducePolymers, LinkedList, Node } from '.'
import { getInput } from '../utils/tests'

describe('part 1', () => {
  describe.each`
    file                   | count
    ${'part-1-case-1.txt'} | ${10}
    ${'index.txt'}         | ${10886}
  `('given $file', async ({ file, count }) => {
    let input: string

    beforeEach(async () => {
      input = await getInput(__dirname, file)
    })

    test('it matches snapshot', () => {
      expect(reducePolymers(input)).toMatchSnapshot()
    })

    test("it's counts match up", () => {
      expect(reducePolymers(input)).toHaveLength(count)
    })
  })
})

describe('LinkedList', () => {
  let list: LinkedList<number>

  beforeEach(() => {
    list = new LinkedList<number>()
      .insertEnd(new Node(10))
      .insertEnd(new Node(15))
      .insertEnd(new Node(7))
      .insertEnd(new Node(11))
      .insertEnd(new Node(12))
      .insertEnd(new Node(13))
      .insertEnd(new Node(14))
      .insertEnd(new Node(6))
  })

  test('it can add items', () => {
    expect(Array.from(list)).toEqual([10, 15, 7, 11, 12, 13, 14, 6])
  })

  test('it can remove items', () => {
    list.remove(list.firstNode!.next!.next!)
    list.remove(list.firstNode!.next!.next!)
    list.remove(list.firstNode!.next!.next!.next!.next!)

    expect(Array.from(list)).toEqual([10, 15, 12, 13, 6])
  })

  test('it preserves previous items', () => {
    list.remove(list.firstNode!.next!.next!)
    list.remove(list.firstNode!.next!.next!)
    list.remove(list.firstNode!.next!.next!.next!.next!)

    const subject = list!.firstNode!.next!.next!.next!.next!.prev!.prev!.prev!
      .prev!.value

    expect(subject).toEqual(10)
  })

  test('it preserves previous items', () => {
    list.remove(list.firstNode!.next!.next!)
    list.remove(list.firstNode!.next!.next!)
    list.remove(list.firstNode!.next!.next!.next!.next!)

    const subject = list!.firstNode!.next!.next!.next!.next!.prev!.prev!.prev!
      .value

    expect(subject).toEqual(15)
  })
})
