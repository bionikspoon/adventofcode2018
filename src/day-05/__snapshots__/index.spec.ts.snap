// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`part 1 given "case-1.txt" it matches snapshot 1`] = `"dabCBAcaDA"`;

exports[`part 1 given "index.txt" it matches snapshot 1`] = `"JQdJkHnAVfxKDoWGYgAiGSnerpIrLrELEONSXNkVAJDxaiBQNNZYQvUgIFZOhYvvLUNyZMVPiESKCEGEMLTjYYLMyIyjaflikQjVgDytPmQWsrsJFeqzfwEMPgadORnYHOuExqWpChqBcxyRewqfbSRqDLnRYxpkrpvCAnDXdFcEwstomkQUUAWEddhbYbfuTZuBIEbgLWynJirXlaZXsoncJhSRpswXSZBggoVlumHQneXTwzKUEItBwPIXTGPgCPScumxUWScWYsDBLnPyeOEzGoSAAzBIXYxhdRVVTBxvnGSneibzgdXGqTmUzAoTigOxGllYdViQioFnEwnzeCQtQtxnevmdvTmwcUOYxzJzemnYKLLRTHJahfgpikNExYOaWPnXbZBETdpRLGSvoeLqdIMeoRPgVxZnnvsjLeOuCDrsiKOOlzvvRiSCYaXhXeKmAgnUnXHnHcEwNTOTSkClwBzkSeWrfMPJuxBUFqgqrUOtLaowrmOpoKGmAJXqaMuVmsbGdvdpMLVlwbuzEqlMkcBppaoViATygUMVRfEbvcAPvBIVeWmmNPUQmzKRsceuflyAbTvzBGrXZMZIvBuCRoKXteUnMPWEuXWHRFTbmlCLgudEBhdLmlXtZbMgUNEXVsFQRLgJtvTnYiypHTXTpOlKmUlSepsJGTeGAFvUvRkrtvMbTwcsRAeRQyOVEWVhVHSDzUiYZAjlDkZbcPKFDrdCEbbfJeRPLvJAzQSACJfhTlbAGZYzGkxAUlCrlmtlMvxTIbQHtLEDZkMzGbsQBgiTIqvyhsPRvvNjFizizlPRpaNgTvMjWVsXJEiWKDlkwdaNzibsrOrvGalKnsmESqykmmeLdLhbtPgFyIoPvbpKLgMOXaSPkRzTZmJLMgQWibANTPgOTNbpgikjzcVPqdYmztemsltfOXwlxLftYDkuIVjCKvJprZQiikFAwRVWeZWFVnOZFTwqZYsReiLbkRQlNDmqMQNxShzeryXlezJvXCNXNFVLJkDuTiDAythsvSHWfoFFSXrkBlaZoYKrjoNYjpnzLtWtsmudaptgqXfuWaInxIepaDlfBaVhzjzdFFkTByWkcKYTiuPSvrBxLbXEcyffZliwdpRTuvrKwzbQrzxtqEvNatJNMkGZmlbQQzyUDkviUAqtOvwQBSlUHjzWZrbcPcxyXtSXrMOTGXNdwsNAuLBFAXuaDUAzfKyCsAOybwKMXsIGlgBBLVKKEDoDAVZXnLMBScxHqGDiEbvgURDsfHgWetELkmqsrGVaHOXeCQPgQwYxMdirtBthYOBOVuZYlNqsCvgKbAJabhqjKhpbkbeHBEsjDwvrBFamRzshghrarQAYWrgpaNvpQHlTunMWZrOhbOLpmDRxzXXgFTVLmJCCkjfyZjssHXLtygUWlZrVhgxRqSBXwHyIRSfMjPkHxriGwAbEtzlojSMuHfwwdkDDFiZegBjGDoFnRyqthSnhkdAsPOqXSxsVSDCpnovUnGeQxYjOyaChQpaPUzpEpbOSClbyAMbuvhgweUfzmCGXiTcBovWItEukJPGKIHDmRLBwyByQXcWApdftMHVIFUMDoJhxLyBBijVeGcAJgZapeLBdwRJbxumNhTnOVkWEKmbAoaDuZMpDNFJDHxBRbNVOfxZpyBfEsfhkZkXvAvFlXrzkAmfbDtoZcVnqWBIRleivCMiyWPUCxTbZkAdSMftsipoTVmEcFGNCjPXwaOgJlWrqxmzDaCRVGDVgiQawKIxxPSiwdlEmlYNMIOjZhqzMRgZauPLYGcshYlVYlUhCMHNutBughknZrVaPEFMYqSdjYnpoQhBkIQxMPPjtVQUqtNIDptpJTSfBwYDmIPyqgZTsecGjlDKibXizDpWHFrTVbNCDseJygloqTApnMSrMbViCRjjILhpolEWeRAEZIYSXZgMGmaZugtXIKxWZPJlVrBsVyWDmWUgWyLazlXifTZpzqbjdhEjEBNSGallbXFuQpmEtvpOgtlGquLFmTSPHjhbYOeMqAjCJlaLqYXOKXSZDWTsegMnRqHFsVavItNBcQjPVpwaYTJKWAVNIgTTOmOWUIcxGptBevZRbxsBihnXDCyKqthrZAXZswwcTRfdcRGxgtbDnnWRAlOfSHJQRWmanyunPNRHKyXJJenuHGJlqzlkSICIRkqvqmeSiHeXGFTKVUbZrFoBErrmZULPsQxYYMfMoLTmOtILmWhmVzGKRfJqpbydPoaHuDGHHpFmbQIrJfuTgUokjOhYMTkIDPowKKoGUOjuTkuXFXQdqaXqmwbYdOLRlFdSPBFeWrhsbbnZwjsFxLPsVVgUyCQFEcWKXZesohUDsjxQSiqmCYNJLQVhNujsOhJqedecaugAroosHPQWxjacKsCvqtiXGneVmPATmxAwDmodhTSWxZTzbgllEnIOMcShCWcSBKMSwLderpCiWJdeVVKtFqzVgCIRBkiJpoMNlzfkdwYaKXsgzIzwfRGwaZTuQEYbVlXmcrXBdimhusEgQOxEaFEFJoISLcbgvQubEzBZCnXUDenYHGXyRkWOTChXrDgyJXqXVDNMLAiXiPGcDvrnkVxeBXPqXlpZYGMxURhIfSkMokTHiUNOJOsleeTVwEnBgViMsRlkoPOtbhaOsRurIRMkVmEiEkzjPuQeJcyUybbrSZZniYSmCdewSfOQWpCmrvnoQtMlXVMJFzQTrNwZcPLvvYCsKlXzSjRRVCEVJQvqipOOYmwoKcqXGLJwwwkUOiodlFawPhXThuyZSgjyjREXKAvoIpcACJcZpsNLMAevpkgeonumpxBFkqGZZxJPoqGoFIopkVkagcSVCUWzAmNCPaOCgaEayEysQlndqlKxIPgIqKbocqnGwbgdHGatsZXPxtVBDYtkZkXRmRKXvNyDnlNkEOkBVCHkHwgmHCbZbbduzCmtXoPXBceVRgSdadXVMARuvTBdvYiLKLRoBNKxcpJMywYoFsEAYpFpMcOhjEkluDNGoCruFKippjUUkohjyNPqEkkuRNlayXPvmTJArqRsqEUmmAuckOMjflrTjTQuyuxPHLOnFIwoJgzlyUIuVVHqiNaoYHflUjEwdEGqYRPPICHLfmpnJXFqpumhuJlxOvmpDmtDqGjLvePBUIzNhVJKhmAxhgXhBYGpkjILeeDqYzvvWDHsvJxkYlOVAEDuNDhLflGDIapkgQpBTJvNNONTfRBSYFmmuWPZmPDDUfeRsnLKbNAGWGnIQmGBjZgYXEIMTMshcWYKaIFgdauWyQtagXYUpwnjBmXDSkyhvOxtxZGTDcASeRgJoUXNqZrteZHYYBFwKLWTZpKVfHmRsdBixTdiAVtErJUwZTRCgmzyVHiCAWOJtrXBYEStgVAmIzJAUhJeupvxwgjMkOcFMBLQneBQvSDpOwzadrMEuyKugvApoSQhjfQZsfcQeXUCHpicwKsfvDFRRRthSpAckVckZgyxlAFDXLYmAqPAlXnGtacHxoyTMKFnelSwiaHHvPYeUQKqqMpvfGsKXQwSoXAxPXHVThdJaxkEQhUwNYBIDaxCSjzPywDRfLGqWuETsigijpZFGxErvrpZfpyfJLoCseAOtUGoCqYWvZATyXvhujxvcWTOHHnkaJCVHrBoLiTuzfRePsWetCIXfafzIHsqsMcsmdmCpvIqSWerpwIxGfdxqfaqgMxkppLYCjOjrwVYMbaxzgKyLPLyAXXdTxsRYPMUmjUQkrFGHxwhOkUFCMZyLyAFWobNoseltUNhZfUinqHrwpGVjhKzhyhRswUPQGUYEdZSISWcnnqjhjvDRoTTbudNPwivNSmBHRvqmjxffxrluDZLoNZssMUPyLWONGCqLYxGzVwDEAwtANlwpszfLrmoEVSbxOKhLwAohLaGpJRFeUaHzSZjgDNVgsGLgtztkULMBQvjNzEhdUIBxpUfHIFHAkQgofujTIKHaZVjCQXOuAKuQBDTmmtFmrXKHQNCqqJSdkIKqmveCNAgvkqLFZJvXNlDRcYfSyvBlQHjXuVhubgTAQTYsMzfwBDRBBeFkqtgzNRXngIWOzHNlNoBrOQLmKhpfWCsabOwKXkYOVdnCCoMfgFJQxrxpbjSpDIKEsnySKtCscAWAwmiLhbHXMyPReTDugMbQXGEFuhstXJSkEaWCCfRunbjHbvebDbObKPlOXsXthBjhIeVGknjXCTohytmsfkUresuLTVUoaPfdmVDCuBInfgvYharmQdqGfkMevHWGHvjbwKMiLqFTRTixfgUnPwbijVlHbISBkNeQkSLjikWgWoBWenPSyaJfuoSZXjEWEQNdIyaJaDodtuWTxWsrVBYXKxQLgvmpazNqrWXvMwbDRpXavdHQKbiPZmJbuwjtndxlBeLZCfNADXVzNDAANgbeVBwIKeVCwbihtoVLssUpnyfgaJEbzjvjsNLgzCtCycrTochYUcetCfnSmqbCIqSkXblpKHHuXgqoNimxGmHYkIchfqAQeerCzvegOVpmaXyPLsUpALZJMsYJtCRTnlWnGARHDKVWneCIVtoSRuuTGEGGXgOwVutinvHPGhKFTvdNSdnvLpevPmtjpzTzCVAGZwQHNruKLgqDzpmwzhAoROmnFZLSyHxHLRSJzyjNRlPmVcDvlDalaJDZxdLkLaHlgLHWNStlvjwdtIsBykUNyKwiYlUFijrgoZGCkncQrqIIHPUMNPbHZGPanzPfXjMTXILjyWEDIAPhPrgdXdUPGWhPDKhDrQaoKwLRxavkPRFFUSxkpboNgZaqeqqiLWSlqORVjOUTGSGlGEjttfYdvebYljXXpSbGGDaOtLcLLPgLtVGlVrasvYBcmhabZBNGokLfPTNZNQLdGYGucERleiDqTGYUWhlKWkigPQaBWNNNLJkgoAHIDRNeWCBeRWDujpmqXHepeyCJFOpMPUUlBtLKMhPdyObRJnvjEnQaXATFlvqGbKcBfnlDaksYEYUkEovKtHCXLuoiaPPJiguVNvNRgLiOircyIeaHUHsBtHHjkgsKJufsjzyIYkuVOemXySnspxHPNohVpRQWkmAfUrsPfEdaRFyxxDuyhUFAyxLvfbPMREseSAncxLAKUDeYecRtlSLRiwXgaYXsbtMrXDCbujdcwBKhWQUvDlwaNXpFLrdVJERNeQNXnUkCuwNoEuMrvsJhqFZmuTYMlZPcDPnYKsjPrjDktilUnIuogOAyktiANMqkgmTLVAdESDltWvROKXloTqUVivbMgNhLvjduSgUjLpdbXkWUBykrUAckpzxkMrvzUdnibVijLAvLiMSLcOSlGisnouutrWeffQBPuXvQBOLXaRKjNsLjmeNrMsRfpXbHKNMMXljLvRExyvRiGczISnWcFcULamWRQRZCOmdtErVCJYpsUeVPKwnLveDWcfsRiQJeoJMOtarIerEtXHRAcVCarhxTeREiRATomjOEjqIrSFCwdEVlNWkpvEuSPyjcvReTDMoczrqrwMAluCfCwNsiZCgIrVYXerVlJLxmmnkhBxPFrSmRnEMJlSnJkrAxlobqVxUpbqFFEwRTUUONSIgLsoClsmIlValJIvBINDuZVRmKXZPKCauRKYbuwKxBDPlJuGsUDJVlHnGmBVIvuQtOLxkorVwTLdseDavltMGKQmnaITKYaoGOUiNuLITKdJRpJSkyNpdCpzLmytUMzfQHjSVRmUeOnWUcKuNxnqEnrejvDRlfPxnAWLdVuqwHkbWCDJUBcdxRmTBSxyAGxWIrlsLTrCEyEdukalXCNasESermpBFVlXYafuHYUdXXYfrADeFpSRuFaMKwqrPvHOnphXPSNsYxMEovUKyiYZJSFUjkSGKJhhTbShuhAEiYCRIoIlGrnVnvUGIjppAIOUlxchTkVOeKuyeySKAdLNFbCkBgQVLftaxAqNeJVNjrBoYDpHmklTbLuupmPofjcYEPEhxQMPJUdwrEbcwEnrdihaOGKjlnnnwbAqpGIKwkLHwuygtQdIELreCUgygDlqnzntpFlKOgnbzBAHMCbyVSARvLgvTlGpllClToAdggBsPxxJLyBEVDyFTTJegLgsgtuoJvroQLswlIQQEQAzGnOBPKXsuffrpKVAXrlWkOAqRdHkdpHwgpuDxDGRpHpaidewYJlixtmJxFpZNApgzhBpnmuphiiQRqCNKcgzOGRJIfuLyIWkYnuKYbSiTDWJVLTsnwhlGLhAlKlDXzdjALAdLVdCvMpLrnJYZjsrlhXhYslzfNMorOaHZWMPZdQGlkURnhqWzgavcZtZPJTMpVEPlVNDsnDVtfkHgphVNITUvWoGxggegtUUrsOTvicENwvkdhragNwLNtrcTjySmjzlaPuSlpYxAMPvoGEVZcREEqaQFHCiKyhMgXMInOQGxUhhkPLBxKsQicBQMsNFcTECuyHCOtRCYcTcZGlnSJVJZBejAGFYNPuSSlvOTHIBWcvEkiWbvEBGnaadnZvxdanFczlEbLXDNTJWUBjMzpIBkqhDVAxPrdBWmVxwRQnZAPMVGlqXkxybvRSwXtwUTDOdAjAYiDnqeweJxzsOUFjAYspNEwbOwGwKIJlsKqEnKbsiBhLvJIBWpNuGFXItrtfQlImkWBJVhgwhVEmKFgQDqMRAHyVGFNibUcdvMDFpAOuvtlUSERuKFSMTYHOtcxJNKgvEiHJbHTxSxoLpkBoBdBEVBhJBNUrFccwAeKsjxTSHUfegxqBmGUdtErpYmxhBHlIMWawaCScTksYNSekidPsJBPXRXqjfGFmOccNDvoyKxkWoBAScwFPHkMlqoRbOnLnhZowiGNxrnZGTQKfEbbrdbWFZmSytqatGBUHvUxJhqLbVYsFyCrdLnxVjzflQKVGancEVMQkiKDsjQQcnqhkxRMfTMMtdbqUkaUoxqcJvzAhkitJUFOGqKahfihFuPXbiuDHeZnJVqbmluKTZTGlgSGvndGJzsZhAuEfrjPgAlHOaWlHkoXBsveOMRlFZSPWLnaTWaedWvZgXylQcgnowlYpumSSznOlzdULRXFFXJMQVrhbMsnVIWpnDUBttOrdVJHJQNNCwsiszDeyugqpuWSrHYHZkHJvgPWRhQNIuFzHnuTLESOnBOwfaYlYzmcfuKoHWXhgfRKquJMumpyrSXtDxxaYlplYkGZXABmyvWRJoJcylPPKXmGQAFQXDFgXiWPREwsQiVPcMDMSCmSQShiZFAFxicTEwSpErFZUtIlObRhvcjAKNhhotwCVXJUHVxYtazVwyQcOguToaEScOljFYPFzPRVReXgfzPJIGISteUwQglFrdWYpZJscXAdibynWuHqeKXAjDHtvhxpXaxOsWqxkSgFVPmQQkquEypVhhAIWsLENfkmtYOXhCATgNxLapQaMylxdfaLXYGzKCvKCaPsHTrrrfdVFSkWCIPhcuxEqCFSzqFJHqsOPaVGUkYUemRDAZWoPdsVqbENqlbmfCoKmJGWXVPUEjHuajZiMavGTseybxRTjowacIhvYZMGcrtzWujReTvaIDtXIbDSrMhFvkPztwlkWfbyyhzETRzQnxuOjGrEsaCdtgzXTXoVHYKsdxMbJNWPuyxGATqYwUADGfiAkywCHSmtmiexyGzJbgMqiNgwganBklNSrEFuddpMzpwUMMfysbrFtnonnVjtbPqGKPAidgLFlHdnUdeavoLyKXjVShdwVVZyQdEEliJKPgybHxGHXaMHkjvHnZiubpEVlJgQdTMdPMVoXLjUHMUPQfxjNPMFlhcippryQgeDWeJuLFhyOAnIQhvvUiuYLZGjOWifNolhpXUYUqtJtRLFJmoKCUaMMueQSrQRajtMVpxYALnrUKKeQpnYJHOKuuJPPIkfURcOgndULKeJHoCmPfPyaeSfOyWYmjPCXknbOrlklIyVDbtVUramvxDADsGrvECbxpOxTMcZUDBBzBchMGWhKhcvbKoeKnLNdYnVxkrMrxKzKTydbvTXpxzSTAghDGBWgNQCOBkQiGpiXkLQDNLqSYeYAeAGcoApcnMaZwucvsCGAKvKPOifOgQOpjXzzgQKfbXPMUNOEGKPVEamlnSPzCjcaCPiOVakxerJYJGszYUHtxHpWAfLDOIouKWWWjlgxQCkOWMyooPIQVqjvecvrrJsZxLkScyVVlpCzWnRtqZfjmvxLmTqONVRMcPwqoFsWEDcMsyINzzsRBBYuYCjEqUpJZKeIeMvKmriRUrSoAHBTopOKLrSmIvGbNeWvtEELSojonuIhtKOmKsFiHruXmgyzPLxQpxbEXvKNRVdCgpIxIalmndvxQxjYGdRxHctowKrYxghyNEduxNczbZeBUqVGBClsiOjfefAeXoqGeSUHMIDbxRCMxLvByeqUtzAWgrFWZiZGSxkAyWDKFZLnmOPjIKbricGvZQfTkrNRvvEDjwIcPREDlWsmkbsCwcHsCmoiNeLLGBZtzXwstHDOMdWaXMtapMvENgxITQVcSkCAJXwqphSOORaGUACEDEQjHoSJUnHvqljnycMQIsqXJSduHOSEzxkwCefqcYuGvvSplXfSJWzNBBSHRwEfbpsDfLrloDyBWMQxAQDqxfxUKtUJougOkkWOpdiKtmyHoJKOuGtUFjRiqBMfPhhgdUhAOpDYBPQjFrkgZvMHwMliToMtlOmFmyyXqSpluzMRRebOfRzBuvktfgxEhIsEMQVQKricisKLZQLjghUNEjjxYkhrnpNUYAMwrqjhsFoLarwNNdBTGXgrCDFrtCWWSzxazRHTQkYcdxNHIbSXBrzVEbTPgXCiuwoMottGinvawkjtyAWPvpJqCbnTiVAvSfhQrNmGEStwdzsxkoxyQlALjcJaQmEoyBHJhpstMflUQgLTGoPVTeMPqUfxBLLAgsnbeJeHDJBQZPztFIxLZAlYwGuwMdwYvSbRvLjpzwXkixTGUzAMgmGzxsyizearEweLOPHliJJrcIvBmRsmNPatQOLGYjESdcnBvtRfhwPdZIxBIkdLJgCEStzGQYpiMdyWbFstjPTPdinTQuqvTJppmXqiKbHqOPNyJDsQymfepAvRzNKHGUbTUnhmcHuLyvLyHSCgylpUAzGrmZQHzJoimnyLMeLDWIspXXikWAqIGvdgvrcAdZMXQRwLjGoAWxpJcngfCeMvtOPISTFmsDaKzBtXcupwYImcVIELribwQNvCzOTdBFMaKZRxLfVaVxKzKHFSeFbYPzXFovnBrbXhdjfndPmzUdAOaBMkewKvoNtHnMUXBjrWDblEPAzGjaCgEvJIbbYlXHjOdmufivhmTFDPawCxqYbYWblrMdhikgpjKUeTiwVObCtIxgcMZFuEWGHVUBmaYBLcsoBPePZupAPqHcAYoJyXqEgNuVONPcdsvSXsxQopSaDKHNsHTQYrNfOdgJbGEzIfddKDWWFhUmsJOLZTeBaWgIRXhKpJmFsriYhWxbsQrXGHvRzLwuGYTlxhSSJzYFJKccjMlvtfGxxZXrdMPloBHoRzwmNUtLhqPVnAPGRwyaqRARHGHSZrMAfbRVWdJSebhEBKBPHkJQHBAjaBkGVcSQnLyzUvoboyHTbTRIDmXyWqGpqcExohAvgRSQMKleTEwGhFSdruGVBeIdgQhXCsbmlNxzvadOdekkvlbbGLgiSxmkWBYoaScYkFZaudAUxafblUanSWDnxgtomRxsTxYXCpCBRzwZJhuLsbqWVoTQauIVKduYZqqBLMzgKmnjTAnVeQTXZRqBZWkRVUtrPDWILzFFYCexBlXbRVspUItykCKwYbtKffDZJZHvAbFLdAPEiXNiAwUFxQGTPADUMSTwTlZNPJynOJRkyOzALbKRxsffOFwTYadItUdKjlvfnxncxVjZELxYREZHsXnqmQMdnLqrKBlIErSyzQWtfzoNvfwzEwvrWafKIIqzRPjVkcJviUKdyTFlXLWxoFTLSMETZMyDQpvCZJKIGPBntoGptnaBIwqGmljMztZrKpsAxomGlkPBVpOiYfGpTBHlDlEMMKYQseMSNkLAgVRoRSBIZnADWKLdkwIejxSvwJmVtGnAPrpLZIZIfJnVVrpSHYVQitIGbqSBgZmKzdelThqBitXVmLTMLRcLuaXKgZyzgaBLtHFjcasqZajVlprEjFBBecDRdfkpCBzKdLJazyIuZdshvHvwevoYqrEarSCWtBmVTRKrVuVfagEtgjSPEsLuMkLoPtxthPYIyNtVTjGlrqfSvxenuGmBzTxLMlDHbeDUGlcLMBtfrhwxUewpmNuETxkOrcUbVizmzxRgbZVtBaYLFUECSrkZMqupnMMwEvibVpaCVBeFrvmuGYtaIvOAPPbCKmLQeZUBWLvlmPDVDgBSMvUmAQxjaMgkOPoMRWOAlTouRQGQfubXUjpmFRwEsKZbWLcKstotnWeChNhxNuNGaMkExHxAycsIrVVZLookISRdcUoElJSVNNzXvGprOEmiDQlEOVsglrPDtebzBxNpwAoyXenKIPGFHAjhtrllkyNMEZjZXyouCWMtVDMVENXTqTqcEZNWeNfOIqIvDyLLgXoGItOaZuMtQgxDGZBIENsgNVXbtvvrDHXyxibZaasOgZeoEYpNlbdSywCswuXMUCspcGpgtxipWbTieukZWtxENqhMULvOGGbzsxWSPrsHjCNOSxNucUnzALxRIjNYwlGBeibUztUFByBHDDewauuqKMOTSWeCfDxdNacVPRKPXyrNldQrsBFQWErYXCbQHcPwQXeUohyNroDAGpmeWFZQEfjSRSwqMpTYdGvJqKILFAJYiYmlyyJtlmegeckseIpvmzYnulVVyHozfiGuVqyznnqbIAXdjavKnxsnoeleRClRiPRENsgIaGygwOdkXFvaNhKjDqj"`;
